### PR TITLE
feat(service): User can import and publish a page from a remote URL

### PR DESCRIPTION
**Pull request by illichso**
## Issue

https://gravitee.atlassian.net/browse/APIM-3548

## Description

As an API publisher

I want to import documentation pages from a remote site like I can with a v2 API

so that I am getting the documentation outside of Gravitee where I am used to managing it.

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-mbydzxavrj.chromatic.com)

### CODE REVIEW

File: gravitee-apim-repository/gravitee-apim-repository-elasticsearch/src/test/java/io/gravitee/repository/elasticsearch/analytics/ElasticsearchAnalyticsRepositoryTest.java [hunk 1]
```diff
@@ -192,7 +192,7 @@ public void testCountWithTwoApis() throws Exception {
         );
 
         assertThat(response).isNotNull();
-        assertThat(response.getCount()).isEqualTo(4);
+        assertThat(response.getCount()).isEqualTo(7);
```
Comment (jourdiw) [inline gravitee-apim-repository/gravitee-apim-repository-elasticsearch/src/test/java/io/gravitee/repository/elasticsearch/analytics/ElasticsearchAnalyticsRepositoryTest.java]: Why did you need to change this test? I don't see the correspondance between the content of the PR and this test.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java [hunk 2]
```diff
@@ -100,6 +102,14 @@ public interface PageMapper {
     @Mapping(target = "uploadDate", source = "createdAt")
     MediaEntity mapMedia(Media media);
 
+    default String map(SourceConfigurationConfiguration value) {
+        try {
+            return mapper.writeValueAsString(value);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException("Failed to serialize configuration", e);
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: Instead of a generic RuntimeException, let's throw a `TechnicalManagementException` (like in line 170)

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        assertEquals(expectedJsonNode, actualJsonNode);
+    }
+
+    @Test
+    @DisplayName("Should throw RuntimeException when serialization fails")
+    void shouldThrowRuntimeExceptionWhenSerializationFails() throws JsonProcessingException {
+        SourceConfigurationConfiguration sourceConfiguration = null;
+
+        var expectedJson = objectMapper.writeValueAsString(null);
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: This test isn't verifying that an exception is thrown.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    void shouldMapSourceConfigurationConfigurationToJsonString() throws JsonProcessingException {
+        var sourceConfiguration = new SourceConfigurationConfiguration();
+        sourceConfiguration.setAutoFetch(true);
+        sourceConfiguration.setUseSystemProxy(true);
+        sourceConfiguration.setUrl("http://example.com");
+
+        var expectedJsonNode = objectMapper.readTree(objectMapper.writeValueAsString(sourceConfiguration));
+        var actualJsonNode = objectMapper.readTree(pageMapper.map(sourceConfiguration));
+
+        assertEquals(expectedJsonNode, actualJsonNode);
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: Instead of comparing the use of the `writeValueAsString` in the mapper vs directly calling the method, WDYT about comparing the expected string? Ex. `"{  \"aFunProperty\": \"value\" }"`. That way we know that the serialization works the way that we want for the database.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/JsonSchemaServiceTest.java [hunk 5]
```diff
@@ -48,10 +48,10 @@ public void shouldRejectInValidJsonConfigurationWithDependencies() throws IOExce
                 "#/ssl/trustStore: required key [content] not found\n" +
                 "#/ssl/trustStore/type: \n" +
                 "#/ssl/trustStore: required key [path] not found\n" +
-                "#/ssl/trustStore: required key [path] not found\n" +
+                "#/ssl/trustStore: required key [content] not found\n" +
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/JsonSchemaServiceTest.java]: Why did this test change?

File: gravitee-apim-repository/gravitee-apim-repository-elasticsearch/src/test/java/io/gravitee/repository/elasticsearch/analytics/ElasticsearchAnalyticsRepositoryTest.java (hunk 1 above)
Comment (illichso) [reply gravitee-apim-repository/gravitee-apim-repository-elasticsearch/src/test/java/io/gravitee/repository/elasticsearch/analytics/ElasticsearchAnalyticsRepositoryTest.java]: It was a flaky test, to be honest. Not sure why but sometimes it passes with 4, sometimes passes with 7.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/JsonSchemaServiceTest.java (hunk 5 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/JsonSchemaServiceTest.java]: This test is flaky.
More times it passes with "content" instead of "path

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java (hunk 2 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: Done

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java (hunk 4 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: Done

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java (hunk 3 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: Done

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/JsonSchemaServiceTest.java (hunk 5 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/JsonSchemaServiceTest.java]: Removed

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java [hunk 6]
```diff
@@ -100,6 +101,13 @@ public interface PageMapper {
     @Mapping(target = "uploadDate", source = "createdAt")
     MediaEntity mapMedia(Media media);
 
+    default String map(Object value) {
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: Seeing as Object is inherited by many types of objects, this may be employed for lots of different mappings 🙈  Is it necessary for the mapper to work?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test
+    @DisplayName("Should map SourceConfigurationConfiguration to JSON string")
+    void shouldMapSourceConfigurationConfigurationToJsonString() throws JsonProcessingException {
+        var sourceConfiguration = new SourceConfiguration();
+        sourceConfiguration.setType("Type1");
+        sourceConfiguration.setConfiguration("Configuration1");
+
+        String expected = """ 
+class SourceConfiguration {
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: Shouldn't it be simply the properties + values of a class? It seems a bit weird that it would include the class name.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java (hunk 6 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: Yeah, because object is in Open API source.configuration so there was a compilation error without this new mapper:
`Can't map property "Object source.configuration" to "String source.configuration". Consider to declare/implement a mapping method: "String map(Object value)".`

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java (hunk 7 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: No, this is how toString is generated in that class from the open API
Screenshot 2024-07-01 at 16 59 09

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java (hunk 6 above)
Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: For this method, let's try to make it as specific as possible to the object it will be receiving. Can you turn this method into a `@Named` method and call it specifically for mapping the `source.configuration` value?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java (hunk 7 above)
Comment (jourdiw) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: Ok, then it needs to be serialized differently than just `.toString()`. With just calling `.toString`, the `class SourceConfiguration {` would be saved in the database for the `configuration` field, which is not ok. Perhaps you can use another method? We do this with lots of other objects, including `Api`.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java (hunk 6 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: Done

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java (hunk 7 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: Done

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java [hunk 8]
```diff
@@ -100,6 +102,22 @@ public interface PageMapper {
     @Mapping(target = "uploadDate", source = "createdAt")
     MediaEntity mapMedia(Media media);
 
+    @Mapping(target = "type", source = "type")
+    @Mapping(target = "configuration", qualifiedByName = "serializeJsonConfiguration")
+    PageSource mapSourceConfigurationToPageSource(SourceConfiguration sourceConfiguration);
+
+    @Named("serializeJsonConfiguration")
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: Seeing as you import the `ConfigurationSerializationMapper`, you are able to use this method named [here](https://github.com/gravitee-io/gravitee-api-management/blob/master/gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/ConfigurationSerializationMapper.java#L37).

So this named method in `PageMapper` is actually a duplicate and can be deleted 😅

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    void setUp() {
+        pageMapper = PageMapper.INSTANCE;
+    }
+
+    @Test
+    @DisplayName("Should map SourceConfigurationConfiguration to JSON string")
+    void shouldMapSourceConfigurationConfigurationToJsonString() throws JsonProcessingException {
+        var sourceConfiguration = new SourceConfiguration();
+        sourceConfiguration.setType("Type1");
+        sourceConfiguration.setConfiguration("Configuration1");
```
Comment (jourdiw) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: This doesn't necessarily look like the configuration that would be sent in this object. It would be best if the test used an object that looks like what the front will be sending to the back. (I believe we have used `LinkedHashMap` in other tests for configuration).

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java (hunk 8 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapper.java]: Removed

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java (hunk 9 above)
Comment (illichso) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/test/java/io/gravitee/rest/api/management/v2/rest/mapper/PageMapperTest.java]: Done

**Review by jourdiw — APPROVED** [review]
